<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Saving and Loading Data from an h5Seurat File</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Saving and Loading Data from an h5Seurat
File</h1>



<p>The h5Seurat file format, based on <a href="https://en.wikipedia.org/wiki/Hierarchical_Data_Format">HDF5</a>,
is specifically designed for the storage and analysis of multi-modal
single-cell and spatially-resolved expression experiments, for example,
from CITE-seq or 10X Visium technologies. It holds all molecular
information and associated metadata, including (for example)
nearest-neighbor graphs, dimensional reduction information, spatial
coordinates and image data, and cluster labels.</p>
<p>This vignette serves as a guide to saving and loading
<code>Seurat</code> objects to h5Seurat files. <strong>This
implementation supports Seurat v5 with full Assay5
compatibility</strong>, automatically handling both legacy Assay and
modern Assay5 objects. For more details about h5Seurat files, please see
the h5Seurat-spec vignette.</p>
<!-- *can't save unserializeable data -->
<div id="saving-a-dataset" class="section level2">
<h2>Saving a dataset</h2>
<p>Saving a <code>Seurat</code> object to an h5Seurat file is a fairly
painless process. All assays, dimensional reductions, spatial images,
and nearest-neighbor graphs are automatically saved as well as extra
metadata such as miscellaneous data, command logs, or cell identity
classes from a <code>Seurat</code> object. To save a <code>Seurat</code>
object, we need the <a href="https://satijalab.org/seurat">Seurat</a>
and srtdisk R packages. Example <code>Seurat</code> objects are
distributed through <a href="https://github.com/satijalab/seurat-data">SeuratData</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(srtdisk)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(SeuratData)</span></code></pre></div>
<p>For this vignette, we’ll use one of the 10X Genomics Visium datasets
from the stxBrain data package. We use this dataset to showcase saving
and loading a dataset with multiple assays, dimensional reductions,
nearest-neighbor graphs, and with spatial image data.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">InstallData</span>(<span class="at">ds =</span> <span class="st">&quot;stxBrain&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">LoadData</span>(<span class="at">ds =</span> <span class="st">&quot;stxBrain&quot;</span>, <span class="at">type =</span> <span class="st">&quot;anterior1&quot;</span>)</span></code></pre></div>
<p>The data loaded is the raw, unprocessed version of the data. In order
to generate the full dataset, we’ll follow the spatial processing
steps.</p>
<details>
<summary>
Processing Steps
</summary>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Update to Seurat v5 format - important for Assay5 compatibility</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">UpdateSeuratObject</span>(brain)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">SCTransform</span>(brain, <span class="at">assay =</span> <span class="st">&quot;Spatial&quot;</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(brain)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(brain, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(brain, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>brain <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(brain, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span></code></pre></div>
</details>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>brain</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co">#&gt; 48721 features across 2696 samples within 2 assays </span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="co">#&gt;  1 other assay present: Spatial</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>As seen, we have a dataset with multiple components to it. Despite
being a complex dataset with multiple parts, saving the dataset is no
problem with nearly all information from the object being written to
disk. Saving an object is as simple as calling
<code>SaveH5Seurat()</code>; minimally, this function takes a
<code>Seurat</code> object and nothing else. Optional arguments are
present for specifying a filename and whether or not you want to
overwrite a preexisting file.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="fu">SaveH5Seurat</span>(brain, <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>On a laptop running Ubuntu 16.04 LTS with an Intel Core i5-2520M
clocked at 2.5 GHz, 16 GB of RAM, and a 512 Gb Samsung Evo SSD, this
process takes ~20 seconds and results in an on-disk file size of roughly
213 Mb.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>size <span class="ot">&lt;-</span> <span class="fu">file.size</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">structure</span>(size, <span class="at">class =</span> <span class="st">&quot;object_size&quot;</span>), <span class="at">units =</span> <span class="st">&quot;Mb&quot;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co">#&gt; 178.6 Mb</span></span></code></pre></div>
<p>An <a href="http://stat.ethz.ch/R-manual/R-devel/library/base/html/readRDS.html">Rds</a>
file with the same object is roughly 200 Mb on disk, though saving the
Rds file took ~42 seconds on the same laptop. Moreover, Rds files are
not easily readable in other languages, such as Python.</p>
</div>
<div id="connecting-to-and-querying-h5seurat-files" class="section level2">
<h2>Connecting to and querying h5Seurat files</h2>
<p>Unlike most data formats, HDF5 files can be connected to and explored
without loading the data into memory. To facilitate this, we’ve built an
<code>h5Seurat</code> object to serve as an interface to h5Seurat files
in R. <code>h5Seurat</code> objects are built off of the
<code>H5File</code> object from the hdf5r package.</p>
<details>
<summary>
<code>h5Seurat</code> objects and R6 classes
</summary>
One thing to note, <code>h5Seurat</code> and <code>H5File</code> objects
are R6 objects. Unlike most R objects (called S3 and S4), and more like
objects in Python, R6 objects are <em>encapsulated</em> objects; this
means that methods are attached directly to the object instead of to a
generic function. <br /> In Seurat, most functions take an object as
input and return an object as output. These functions actually run
differently depending on the class of the object passed to them. For
example, <code>RunUMAP</code> has 4 different modes of operation,
depending on the type of object that’s passed to it. One can see which
objects trigger different routines by using the <a href="https://stat.ethz.ch/R-manual/R-patched/library/utils/html/methods.html"><code>methods</code></a>
function. Functions that change behavior are known as “generics” and the
exact implementations are known as “methods”; these methods are
associated with the generic instead of with the object itself. <br /> R6
objects, however, have their methods attached directly to the object.
Calling an R6 method is done similarly to data access in R’s S3 and S4
object system: using the <code>$</code> operator. For example, creating
a new Seurat object is done with <code>CreateSeuratObject</code> or
<code>new(Class = &quot;Seurat&quot;)</code> (for advanced users), while
initializing a new <code>h5Seurat</code> object is done with
<code>h5Seurat$new()</code> <br /> For more details about R6 objects,
please see the <a href="https://r6.r-lib.org/">R6 website and
documentation</a>
</details>
<p>Connecting to an h5Seurat file is as simple as instantiating an
<code>h5Seurat</code> object.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>hfile <span class="ot">&lt;-</span> <span class="fu">Connect</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>hfile</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">#&gt; Class: h5Seurat</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co">#&gt; Filename: /Users/miana/Documents/GitHub/srtdisk/vignettes/anterior1.h5Seurat</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="co">#&gt; Access type: H5F_ACC_RDONLY</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co">#&gt; Attributes: version, project, active.assay</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="co">#&gt; Listing:</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="co">#&gt;          name    obj_type dataset.dims dataset.type_class</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="co">#&gt;  active.ident   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="co">#&gt;        assays   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co">#&gt;    cell.names H5I_DATASET         2696         H5T_STRING</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="co">#&gt;      commands   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="co">#&gt;        graphs   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co">#&gt;        images   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co">#&gt;     meta.data   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co">#&gt;          misc   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="co">#&gt;     neighbors   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="co">#&gt;    reductions   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="co">#&gt; &lt; Printed 10, out of 11&gt;</span></span></code></pre></div>
<p>As seen, the h5Seurat file is structured similarly to a
<code>Seurat</code> object, with different HDF5 groups sharing the names
of slots in a <code>Seurat</code> object. However, it’s difficult to
glean what data is present in this dataset similar to calling a
<code>Seurat</code> object in the R console. To get around this, we’ve
created an <code>index</code> method for <code>h5Seurat</code> objects;
this method creates a summary of the data stored within the h5Seurat
object. As <code>Seurat</code> objects are organized around the assay
data, this h5Seurat index showcases the data grouped by assay.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>hfile<span class="sc">$</span><span class="fu">index</span>()</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co">#&gt; Data for assay SCT★ (default assay)</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="co">#&gt;    counts      data    scale.data</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">#&gt;      ✔          ✔          ✔          ✔     </span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co">#&gt; Dimensional reductions:</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="co">#&gt;         Embeddings  Loadings  Projected  JackStraw </span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">#&gt;  pca:       ✔          ✔          ✖          ✖     </span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="co">#&gt;  umap:      ✔          ✖          ✖          ✖     </span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co">#&gt; Graphs:</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="co">#&gt;  ─ SCT_nn</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">#&gt;  ─ SCT_snn</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="co">#&gt; Data for assay Spatial</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="co">#&gt;    counts      data    scale.data</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="co">#&gt;      ✔          ✖          ✖     </span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co">#&gt; Images:</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="co">#&gt;  ─ anterior1</span></span></code></pre></div>
<p>First we get a breakdown of what slots are filled within each assay,
followed by a table of dimensional reduction information. This table
shows which bits of information (eg. cell embeddings, feature loadings,
JackStraw data) are present. these tables, we get a list of
nearest-neighbor graphs and spatial image data. This way, we can see
what data gets loaded on a per-assay basis as is required by Seurat.</p>
<p>To explore an h5Seurat file deeper, we can use the double bracket
<code>[[</code> operator to explore various aspects of the dataset. The
double bracket <code>[[</code> operator takes a UNIX-style path
comprised of dataset names.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>hfile[[<span class="st">&quot;assays&quot;</span>]]</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co">#&gt; Class: H5Group</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#&gt; Filename: /Users/miana/Documents/GitHub/srtdisk/vignettes/anterior1.h5Seurat</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#&gt; Group: /assays</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#&gt; Listing:</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#&gt;     name  obj_type dataset.dims dataset.type_class</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#&gt;      SCT H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#&gt;  Spatial H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>hfile[[<span class="st">&quot;assays/SCT&quot;</span>]]</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co">#&gt; Class: H5Group</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co">#&gt; Filename: /Users/miana/Documents/GitHub/srtdisk/vignettes/anterior1.h5Seurat</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co">#&gt; Group: /assays/SCT</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#&gt; Attributes: key, s4class</span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#&gt; Listing:</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#&gt;               name    obj_type dataset.dims dataset.type_class</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co">#&gt;      SCTModel.list   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co">#&gt;             counts   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="co">#&gt;               data   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co">#&gt;           features H5I_DATASET        17668         H5T_STRING</span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co">#&gt;               misc   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#&gt;         scale.data H5I_DATASET  3000 x 2696          H5T_FLOAT</span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="co">#&gt;    scaled.features H5I_DATASET         3000         H5T_STRING</span></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="co">#&gt;  variable.features H5I_DATASET         3000         H5T_STRING</span></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a>hfile[[<span class="st">&quot;reductions&quot;</span>]]</span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a><span class="co">#&gt; Class: H5Group</span></span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a><span class="co">#&gt; Filename: /Users/miana/Documents/GitHub/srtdisk/vignettes/anterior1.h5Seurat</span></span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a><span class="co">#&gt; Group: /reductions</span></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a><span class="co">#&gt; Listing:</span></span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a><span class="co">#&gt;  name  obj_type dataset.dims dataset.type_class</span></span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a><span class="co">#&gt;   pca H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a><span class="co">#&gt;  umap H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>hfile[[<span class="st">&quot;reductions/umap&quot;</span>]]</span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a><span class="co">#&gt; Class: H5Group</span></span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a><span class="co">#&gt; Filename: /Users/miana/Documents/GitHub/srtdisk/vignettes/anterior1.h5Seurat</span></span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a><span class="co">#&gt; Group: /reductions/umap</span></span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a><span class="co">#&gt; Attributes: active.assay, key, global</span></span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a><span class="co">#&gt; Listing:</span></span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a><span class="co">#&gt;             name    obj_type dataset.dims dataset.type_class</span></span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a><span class="co">#&gt;  cell.embeddings H5I_DATASET     2696 x 2          H5T_FLOAT</span></span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a><span class="co">#&gt;             misc   H5I_GROUP         &lt;NA&gt;               &lt;NA&gt;</span></span></code></pre></div>
<p>When finished exploring an h5Seurat file, remember to close the
connection. Because we’re working with file on disk directly rather than
loading it into memory, we need to close it to prevent file corruption.
You can also open the file in read-only mode (<code>mode = &quot;r&quot;</code>)
to help alleviate file corruption, though it’s still a good habit to
close the h5Seurat file when done working with it.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a>hfile<span class="sc">$</span><span class="fu">close_all</span>()</span></code></pre></div>
</div>
<div id="loading-datasets" class="section level2">
<h2>Loading datasets</h2>
<p>Reading data from an h5Seurat file is as simple as calling
<code>LoadH5Seurat()</code>; by default, it loads the entire object into
memory.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>brain2</span></code></pre></div>
<p>However, there are situations in which loading an entire
<code>Seurat</code> object is not desirable. As such, we can leverage
the HDF5 format and load only parts of a dataset at a time.
<code>LoadH5Seurat</code> makes use of assay association to limit the
data loaded. In <code>Seurat</code> objects, all dimensional reduction
information, nearest-neighbor graphs, and spatial image data have an
assay they “belong” to (see <code>DefaultAssay()</code> in the Seurat
package for more details). If only certain assays are requested, then
only the object associated with those assays are loaded.</p>
<p>There are four main parameters for controlling data loading. The
first is the <code>assays</code> parameter; this parameter controls
which assays are loaded and which slots of each assay are loaded. The
simplest level of control is specifying the assays to load. For our
brain dataset, we can choose from either “SCT” or “Spatial”; passing one
of these will load the entire assay object for the assay specified.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, <span class="at">assays =</span> <span class="st">&quot;SCT&quot;</span>)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>brain2</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#&gt; 17668 features across 2696 samples within 1 assay </span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>We can also choose the slots to load; the slots available are
“counts” for the raw expression data, “data” for the normalized
expression data, or “scale.data” for the scaled expression data.
Specifying slots instead of assays will load the desired slots from all
assays that have the requested slots. When specifying slots, one of
either “counts” or “data” <strong>must</strong> be specified as the
<code>Seurat</code> object uses these slots to control dataset
dimensionality information.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, <span class="at">assays =</span> <span class="st">&quot;data&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>brain2</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co">#&gt; 17668 features across 2696 samples within 1 assay </span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>For more fine-tuned control, the <code>assays</code> parameter can
also take a named list or vector, where the names are the names of the
assays to load and the values are the slots to load.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">SCT =</span> <span class="fu">c</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;scale.data&quot;</span>),</span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>    <span class="at">Spatial =</span> <span class="st">&quot;counts&quot;</span>))</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>brain2</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co">#&gt; 48721 features across 2696 samples within 2 assays </span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="co">#&gt;  2 layers present: counts, scale.data</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co">#&gt;  1 other assay present: Spatial</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>Finally, passing <code>NULL</code> to <code>assays</code> (the
default behavior) loads all assays and all slots.</p>
<p>The second of the main parameters is the <code>reductions</code>
parameter; this parameter controls which dimensional reductions are
loaded. As dimensional reductions are tied to assays, the data request
needs to be either associated to a loaded assay or marked as global to
be loaded (see details below). For example, trying to load the “pca”
reduction with the “Spatial” assay won’t work as the “pca” reduction is
associated with the “SCT” assay and not marked as global.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, <span class="at">assays =</span> <span class="st">&quot;Spatial&quot;</span>, <span class="at">reductions =</span> <span class="st">&quot;pca&quot;</span>)</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>brain2</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="co">#&gt; 31053 features across 2696 samples within 1 assay </span></span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a><span class="co">#&gt; Active assay: Spatial (31053 features, 0 variable features)</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>There are three special values the <code>reductions</code> parameter
can take: <code>NULL</code> for all dimensional reductions that can be
loaded (the default behavior), <code>NA</code> for <em>global</em>
dimensional reductions only, or <code>FALSE</code> for no dimensional
reduction information.</p>
<p>The <code>graphs</code> parameter is the third main parameter; this
parameter controls which nearest-neighbor graphs to load. Just like
dimensional reduction information, nearest-neighbor graphs are tied to
assays, and thus are only loaded when their associated assay is loaded
as well. There are two special values the <code>graphs</code> parameter
can take: <code>NULL</code> for all graphs that can be loaded (the
default behavior) or <code>FALSE</code> for no nearest-neighbor
graphs.</p>
<p>The final main parameter is the <code>images</code> parameter; this
parameter controls which spatial image data is loaded. All spatial image
data are marked global by default, so they are loaded whether or not
their associated assays are loaded as well. The <code>images</code>
parameter has three special values: <code>NULL</code> for all spatial
image data (the default), <code>NA</code> for <em>global</em> spatial
image data (typically the same as <code>NULL</code>), or
<code>FALSE</code> for no spatial image data.</p>
<p>With these four parameters, there is a lot of customization for
loading <code>Seurat</code> objects from h5Seurat files. For example,
the following will load the “data” slot from the “Spatial” assay, the
“data” and “scale.data” slots from the “SCT” assay, <em>global</em>
dimensional reductions, none of the nearest neighbor graphs, and all
spatial images.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">SCT =</span> <span class="fu">c</span>(<span class="st">&quot;data&quot;</span>, <span class="st">&quot;scale.data&quot;</span>),</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>    <span class="at">Spatial =</span> <span class="st">&quot;counts&quot;</span>), <span class="at">reductions =</span> <span class="cn">NA</span>, <span class="at">graphs =</span> <span class="cn">FALSE</span>, <span class="at">images =</span> <span class="cn">NULL</span>)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>brain2</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co">#&gt; 48721 features across 2696 samples within 2 assays </span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#&gt;  2 layers present: counts, scale.data</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co">#&gt;  1 other assay present: Spatial</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a><span class="co">#&gt;  1 dimensional reduction calculated: umap</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>In addition, there are four secondary parameters to
<code>LoadH5Seurat</code>: <code>meta.data</code>,
<code>commands</code>, <code>misc</code>, and <code>tools</code>; these
all take simple <code>TRUE</code>/<code>FALSE</code> values to control
the loading of cell-level metadata, command logs, miscellaneous
information, or tool-specific results, respectively.</p>
<details>
<summary>
Global objects
</summary>
The concept of global objects in Seurat is designed as an extension to
the assay-centric nature of <code>Seurat</code> objects. In Seurat, each
assay is considered to be one experiment or measurement of data for a
common group of cells. These assays are then used to generate working
summaries, such as reduced dimension space or nearest-neighbor graphs.
Generally, if an assay is removed from an object, the working summaries
are of little use so they get removed as well. <br /> However, there are
some instances in which these working summaries are useful outside the
context of their assay. For example, some reduced representations of the
data such as tSNE or UMAP are useful for visualization regardless of the
the assay. As such, certain reduced representations and all spatial
image data are marked as <em>global</em>, allowing them to persist as
useful visualization contexts without their associated assay and large
expression matrices being present. <br /> For more details about global
objects, please see the documentation for <a href="https://rdrr.io/cran/Seurat/man/IsGlobal.html">globality in
Seurat</a>
</details>
<p>Partial loading of datasets is an excellent way to limit memory usage
and prevent the loading of massive datasets into memory. However, there
can be instances in which a partial dataset was loaded, but then needs
to be expanded with additional data from the h5Seurat file. Instead of
redoing the partial load, we can make use of <code>AppendData()</code>
to add additional objects from an h5Seurat file to an already-loaded
<code>Seurat</code> object. To show how this works, we’ll start off with
by loading just the “data” slot from the “SCT” assay as well as all
spatial image data, but not load any dimensional reduction information
or nearest-neighbor graphs.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, <span class="at">assays =</span> <span class="fu">c</span>(<span class="at">SCT =</span> <span class="st">&quot;data&quot;</span>), <span class="at">reductions =</span> <span class="cn">FALSE</span>,</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>    <span class="at">graphs =</span> <span class="cn">FALSE</span>, <span class="at">images =</span> <span class="cn">NULL</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>brain2</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="co">#&gt; 17668 features across 2696 samples within 1 assay </span></span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p><code>AppendData</code> takes the h5Seurat file, the
<code>Seurat</code> object generated from <code>LoadH5Seurat</code> and
uses the four main paramters from <code>LoadH5Seurat</code>
(<code>assays</code>, <code>reductions</code>, <code>graphs</code>, and
<code>images</code>). These parameters are used in the same way as
<code>LoadH5Seurat</code> with one exception: <code>assays</code> can
now take <code>FALSE</code> as a value. By passing <code>FALSE</code>,
we prevent other assay information from being loaded; this is useful if
we only want to add other bits of data to our <code>Seurat</code>
object. For example, we can choose to add only <em>global</em>
dimensional reductions to the already existing <code>Seurat</code>
object.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">AppendData</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, brain2, <span class="at">assays =</span> <span class="cn">FALSE</span>, <span class="at">reductions =</span> <span class="cn">NA</span>,</span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>    <span class="at">graphs =</span> <span class="cn">FALSE</span>, <span class="at">images =</span> <span class="cn">NULL</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>brain2</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co">#&gt; 17668 features across 2696 samples within 1 assay </span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a><span class="co">#&gt;  1 dimensional reduction calculated: umap</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>The only limits to the number of times <code>AppendData</code> can be
run is when the <code>h5Seurat</code> file has run out of data not
present in the <code>Seurat</code> object. Otherwise, it can be run
multiple times, adding new bits of data to our <code>Seurat</code>
object. Here, we fill out the rest of the “SCT” assay, but load no other
information</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">AppendData</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, brain2, <span class="at">assays =</span> <span class="st">&quot;SCT&quot;</span>, <span class="at">reductions =</span> <span class="cn">FALSE</span>,</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>    <span class="at">graphs =</span> <span class="cn">FALSE</span>, <span class="at">images =</span> <span class="cn">FALSE</span>)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>brain2</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="co">#&gt; 17668 features across 2696 samples within 1 assay </span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="co">#&gt;  1 dimensional reduction calculated: umap</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
<p>If we want to perform a “full append” (loading all bits of data of a
<code>Seurat</code> object from an h5Seurat file), we can set the four
parameters to <code>NULL</code>, which happens to be the default values
for these parmaters. This loads the rest of the <code>Seurat</code>
object from the h5Seurat file into memory.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>brain2 <span class="ot">&lt;-</span> <span class="fu">AppendData</span>(<span class="st">&quot;anterior1.h5Seurat&quot;</span>, brain2)</span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>brain2</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="co">#&gt; An object of class Seurat </span></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="co">#&gt; 48721 features across 2696 samples within 2 assays </span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="co">#&gt; Active assay: SCT (17668 features, 3000 variable features)</span></span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a><span class="co">#&gt;  1 layer present: counts</span></span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a><span class="co">#&gt;  1 other assay present: Spatial</span></span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a><span class="co">#&gt;  2 dimensional reductions calculated: umap, pca</span></span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a><span class="co">#&gt;  1 spatial field of view present: anterior1</span></span></code></pre></div>
</div>
<div id="working-with-external-visium-datasets" class="section level2">
<h2>Working with external Visium datasets</h2>
<p>While <code>SeuratData</code> provides curated examples, many Visium
datasets are shared as AnnData (<code>.h5ad</code>) files through
resources such as <a href="https://cellxgene.cziscience.com">cellxgene</a>. The
<code>srtdisk</code> package can download these files, convert them to
h5Seurat, and reuse the partial loading patterns demonstrated above.</p>
<div id="downloading-the-heart-visium-dataset" class="section level3">
<h3>Downloading the Heart Visium dataset</h3>
<p>The CZ Biohub Spatial Atlas heart dataset is available as an
<code>.h5ad</code> file. The snippet below downloads the data to a
user-defined cache directory; the chunk is not evaluated during vignette
rendering to avoid large downloads.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>heart_url <span class="ot">&lt;-</span> <span class="st">&quot;https://datasets.cellxgene.cziscience.com/3c84ef48-2cc2-4327-90a1-4f4cb57de168.h5ad&quot;</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>cache_dir <span class="ot">&lt;-</span> fs<span class="sc">::</span><span class="fu">path_home_r</span>(<span class="st">&quot;.cache&quot;</span>, <span class="st">&quot;srtdisk&quot;</span>, <span class="st">&quot;data&quot;</span>)</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>fs<span class="sc">::</span><span class="fu">dir_create</span>(cache_dir, <span class="at">recurse =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>heart_h5ad <span class="ot">&lt;-</span> fs<span class="sc">::</span><span class="fu">path</span>(cache_dir, <span class="st">&quot;heart_visium.h5ad&quot;</span>)</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a></span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span>fs<span class="sc">::</span><span class="fu">file_exists</span>(heart_h5ad)) {</span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a>    <span class="fu">download.file</span>(heart_url, <span class="at">destfile =</span> heart_h5ad, <span class="at">mode =</span> <span class="st">&quot;wb&quot;</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div id="converting-to-h5seurat" class="section level3">
<h3>Converting to h5Seurat</h3>
<p>Once the AnnData file is present, converting and loading it follows
the same pattern as earlier sections. The conversion is disabled during
knitting because it requires Python dependencies, but the code is ready
to copy and run in an interactive session.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a>heart_h5seurat <span class="ot">&lt;-</span> fs<span class="sc">::</span><span class="fu">path</span>(cache_dir, <span class="st">&quot;heart_visium.h5seurat&quot;</span>)</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="fu">Convert</span>(heart_h5ad, <span class="at">dest =</span> heart_h5seurat, <span class="at">overwrite =</span> <span class="cn">TRUE</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>heart <span class="ot">&lt;-</span> <span class="fu">LoadH5Seurat</span>(heart_h5seurat, <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">Spatial =</span> <span class="st">&quot;counts&quot;</span>), <span class="at">images =</span> <span class="cn">NULL</span>,</span>
<span id="cb22-6"><a href="#cb22-6" tabindex="-1"></a>    <span class="at">reductions =</span> <span class="cn">NA</span>, <span class="at">graphs =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-7"><a href="#cb22-7" tabindex="-1"></a>heart</span></code></pre></div>
<p>You can now explore and append additional data just as we did with
the brain example, taking advantage of on-disk storage to manage large
spatial datasets.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
